### PR TITLE
chore(gatsby-link): Migrate gatsby-link to TypeScript

### PR DESCRIPTION
**Pull request by Eyas**
## Description

This PR migrates `packages/gatsby-link` to TypeScript. Highlights:

- Replaces index.d.ts with actual typings generated from source .ts
- Tests are in .ts[x]? too
- Move package outputs to `dist/`
- Typings should be 1:1 compatible.

## Related Issues

Part of #21995.

### CODE REVIEW

**Comment by Eyas** [comment]
e2e tests are falining with something like:

```
Uncaught SyntaxError: Unexpected token <
```

Not sure if there's another place I'd need to propagate some babel settings to get this working. I'm assuming this is due to .tsx

**Comment by Eyas** [comment]
I just saw #22027 already attempts this. Sorry for duplicating this work.

Note that I also remove index.d.ts but added a typings generation step to introduce a compatible one. Happy to work on this on either PR

File: packages/gatsby-link/src/internal.d.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  const ___loader: {
+    enqueue: (...args: unknown[]) => void
+    hovering: (arg: unknown) => void
+  }
+
+  interface Window {
+    ___navigate(arg: unknown, options: unknown): void
+    ___push(arg: unknown): void
+    ___replace(arg: unknown): void
```
Comment (blainekasten) [inline packages/gatsby-link/src/internal.d.ts]: why are these considered `unknown`? We know these types don't we?

File: packages/gatsby-link/tsconfig.json [hunk 2]
```diff
@@ -1,3 +1,4 @@
 {
-  "extends": "../../tsconfig.json"
+  "extends": "../../tsconfig.json",
+  "exclude": ["src/__tests__", "dist/"]
```
Comment (blainekasten) [inline packages/gatsby-link/tsconfig.json]: why is this required? I don't think the other projects needed this exclude.

File: packages/gatsby-link/src/index.tsx [hunk 3]
```diff
@@ -1,12 +1,48 @@
...
 export { parsePath }
 
-export function withPrefix(path) {
+export interface IGatsbyLinkState {
+  IOSupported: boolean
+}
+
+// eslint-disable-next-line @typescript-eslint/interface-name-prefix
+export interface GatsbyLinkProps<TState = unknown>
+  extends React.PropsWithoutRef<LinkProps<TState>> {
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: what does `React.PropsWithoutRef` give us here?

File: packages/gatsby-link/src/index.tsx [hunk 4]
```diff
@@ -1,12 +1,48 @@
...
 import { parsePath } from "./parse-path"
 
 export { parsePath }
 
-export function withPrefix(path) {
+export interface IGatsbyLinkState {
+  IOSupported: boolean
+}
+
+// eslint-disable-next-line @typescript-eslint/interface-name-prefix
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: why are we doing this eslint disable? let's just name it `IGatsbyLinkProps`

File: packages/gatsby-link/src/index.tsx [hunk 5]
```diff
@@ -1,12 +1,48 @@
...
+export interface GatsbyLinkProps<TState = unknown>
+  extends React.PropsWithoutRef<LinkProps<TState>> {
+  /** A class to apply when this Link is active */
+  activeClassName?: string
+
+  /** Inline styles for when this Link is active */
+  activeStyle?: {}
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: does react ship a style type that we can use for this?

File: packages/gatsby-link/src/index.tsx [hunk 6]
```diff
@@ -15,11 +51,22 @@ export function withPrefix(path) {
...
 }
 
-function normalizePath(path) {
+/**
+ * Sometimes you need to navigate to pages programmatically, such as during form submissions. In these
+ * cases, `Link` won’t work.
+ */
+export const navigate: NavigateFn = async (
+  to: string | number,
+  options?: NavigateOptions<{}>
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: should `{}` be a passed in generic? I'm assuming it's for forwarding state

File: packages/gatsby-link/src/index.tsx [hunk 7]
```diff
@@ -88,9 +161,9 @@ class GatsbyLink extends React.Component {
     instance.disconnect()
   }
 
-  handleRef(ref) {
-    if (this.props.innerRef && this.props.innerRef.hasOwnProperty(`current`)) {
-      this.props.innerRef.current = ref
+  handleRef: React.Ref<HTMLAnchorElement> = (ref: HTMLAnchorElement) => {
+    if (refIsObject(this.props.innerRef)) {
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: I don't see the benefit of extracting this as it's not reused. Can we keep the condition explicit here?

File: packages/gatsby-link/src/index.tsx [hunk 8]
```diff
@@ -88,9 +161,9 @@ class GatsbyLink extends React.Component {
...
+      asMutable(this.props.innerRef).current = ref
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: I am severely confused by this entire `handlRef` function... Like we shouldn't need to do any of this. We should just be passing `innerRef={this.props.ref}` in the below render. Then move the createIntersectionObserver to a componentDidUpdate or something. 

I also don't see when `this.props.innerRef` would be a function. Or why it is.. idk (maybe it's just late and I'm tired 😆 )

File: packages/gatsby-link/src/index.tsx [hunk 9]
```diff
@@ -112,23 +188,26 @@ class GatsbyLink extends React.Component {
         style: { ...this.props.style, ...this.props.activeStyle },
       }
     }
-    return null
+
+    // TODO: Reach typings state that getProps cannot return null.
+    return (null as unknown) as {}
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: if that's true should we just be returning `{}` ?

File: packages/gatsby-link/src/index.tsx [hunk 10]
```diff
@@ -112,23 +188,26 @@ class GatsbyLink extends React.Component {
...
-  render() {
+  render(): JSX.Element {
     const {
       to,
       getProps = this.defaultGetProps,
       onClick,
       onMouseEnter,
-      /* eslint-disable no-unused-vars */
+      /* eslint-disable @typescript-eslint/no-unused-vars */
+      // Explicitly destructure these away so our "rest" params don't include them.
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: s/don't/doesn't

File: packages/gatsby-link/src/index.tsx [hunk 11]
```diff
@@ -146,14 +225,14 @@ class GatsbyLink extends React.Component {
         to={prefixedTo}
         state={state}
         getProps={getProps}
-        innerRef={this.handleRef}
-        onMouseEnter={e => {
+        innerRef={this.handleRef as React.Ref<HTMLAnchorElement>}
+        onMouseEnter={(e): void => {
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: add type `e`

File: packages/gatsby-link/src/index.tsx [hunk 12]
```diff
@@ -146,14 +225,14 @@ class GatsbyLink extends React.Component {
...
           if (onMouseEnter) {
             onMouseEnter(e)
           }
           ___loader.hovering(parsePath(to).pathname)
         }}
-        onClick={e => {
+        onClick={(e): boolean => {
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: add type `e`

File: packages/gatsby-link/src/index.tsx [hunk 13]
```diff
@@ -182,39 +261,55 @@ class GatsbyLink extends React.Component {
...
 
-export default React.forwardRef((props, ref) => (
-  <GatsbyLink innerRef={ref} {...props} />
+/**
+ * This component is intended _only_ for links to pages handled by Gatsby. For links to pages on other
+ * domains or pages on the same domain not handled by the current Gatsby site, use the normal `<a>` element.
+ */
+const GatsbyLink = React.forwardRef<
+  HTMLAnchorElement,
+  GatsbyLinkProps<unknown>
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: don't we need to pass in a generic here for it to ever resolve to anything in the actual `GatsbyLinkInternal` functions?

Comment (blainekasten) [reply packages/gatsby-link/src/index.tsx]: I think since we are explicitly typing the arguments there isn't a benefit to passing in the generics

File: packages/gatsby-link/src/index.tsx [hunk 14]
```diff
@@ -182,39 +261,55 @@ class GatsbyLink extends React.Component {
...
+>((props: GatsbyLinkProps<unknown>, ref: React.Ref<HTMLAnchorElement>) => (
+  <GatsbyLinkInternal innerRef={ref} {...props} />
 ))
+type GatsbyLink<TState = unknown> = React.Component<
+  GatsbyLinkProps<TState>,
+  IGatsbyLinkState
+>
```
Comment (blainekasten) [inline packages/gatsby-link/src/index.tsx]: does this not cause problems declaring the type as the same as the const above?

File: packages/gatsby-link/src/internal.d.ts (hunk 1 above)
Comment (Eyas) [reply packages/gatsby-link/src/internal.d.ts]: Yeah honestly I'm not sure (and I don't know where they come from), so pointers would be appreciated!

File: packages/gatsby-link/tsconfig.json (hunk 2 above)
Comment (Eyas) [reply packages/gatsby-link/tsconfig.json]: I saw this pattern in `packages/gatsby-core-utils/tsconfig.json`

Basically if there's a "typegen" step in NPM, tsc will generate .d.ts files for tests. This causes two problems:

1. Looks like jest tries to run the output ".d.ts" files as tests and complains there are no test cases
2. We wouldn't want __tests__ to be exportable types.

File: packages/gatsby-link/src/index.tsx (hunk 3 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: It gives us all of `LinkProps<TState>` except for the "ref" property.

It isn't entirely clear to me how the @reach/router `LinkProps` type ends up getting a ref type (somewhere from react's core typings), but it ends up having the property:

```
    ref: Link<TState>
```

which ends up getting confused with Gatsby's view of the ref: an HTMLAnchorElement.

I can dig more into the disconnect.

File: packages/gatsby-link/src/index.tsx (hunk 4 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: `GatsbyLinkProps` was previously exported in index.d.ts, so this was my attempt at keeping the new generated index.d.ts compatible.

Alternatively, I can name this IGatsbyLinkProps and export a deprecated typedef called GatsbyLinkProps? or what do you think?

File: packages/gatsby-link/src/index.tsx (hunk 5 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: I couldn't find it, but I'm not sure where these come from in js. Do you know what the corresponding npm package that gives us these? Doesn't seem to be in react core

File: packages/gatsby-link/src/index.tsx (hunk 7 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: `React.Ref<T>` is defined roughly as `React.RefFunction<T> | React.RefObject<T>`, but the check above doesn't help typescript narrow to `RefObject` in this branch  (or a function in the else branch). So the benefit of factoring out is to explicitly annotate the type assertion return type.

Alternatively, I can leave it as is and add a cast on both `if` and `else` branches.

File: packages/gatsby-link/src/index.tsx (hunk 8 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: Yeah so Ref can be either a function or an object, so I think someone passing GatsbyLink a function ref would do it? Not sure.

Regarding your first point... yeah... the logic does feel convoluted. I was hoping to avoid changing it from what it was initially, so that I don't accidentally change the behavior, but if someone more intimiately familiar with the original intent can vet a refactor, then it might make sense!

File: packages/gatsby-link/src/index.tsx (hunk 13 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: In TS var/const/let variables cannot be declared as generic, so there's no way to parameterize the `const GatsbyLink`. The `type GatsbyLink` below achieves parameterizing this through [TypeScript declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces).

The mean reason I'm doing this is because the old index.d.ts declared GatsbyLink as

```
export class GastbyLink<TState> extends React.Component<....>
```

File: packages/gatsby-link/src/index.tsx (hunk 14 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: See note above on TypeScript declaration merging. It doesn't quite cause answers, but it's also not the prettiest. But we're in a bit of a bind because of the previous index.d.ts declaring it this way and consts not allowing generic types.

File: packages/gatsby-link/src/index.tsx (hunk 3 above)
Comment (Eyas) [reply packages/gatsby-link/src/index.tsx]: @blainekasten  Ah, here's the problem:

the reach router LinkProps incorrectly includes a "ref" type:

```
type RefType = LinkProps<unknown>["ref"];
```

which is: `React.Ref<HTMLAnchorElement>`.

Note that in React, `ref` isn't supposed to be recorded on `props` per se, but on the actual component's intrinsic types.

But @reach/router typings define LinkProps to extend AnchorProps, which eventually include a "ref". While this is the correct ref type on @reach/router's `Link`, it shouldn't be on props.

Anyway, so **why** does "ref" existing become an issue? Because of how `GatsbyLink` is defined:

```
const GatsbyLink = React.forwardRef((props, ref) =>
  <GatsbyLinkInternal innerRef={ref} {...props} />
));
```

the problem is that `...props` is of type `GatsbyLinkProps`, which transitively includes "ref".

BUT `<GatsbyLinkInternal ref=...>` comes from `IntrinsicClassAtributes<GatsbyLinkInternal>` which is `React.Ref<GatsbyLinkInternal>`.... not at all what we want.

**Comment by Eyas** [comment]
@blainekasten thanks for the review. Just resolved some of these comments and replied to the rest.

Note that I'm also seeing a problem with cypress tests not recognizing some ts syntax?

> e2e tests are failing with something like:
> 
> ```
> Uncaught SyntaxError: Unexpected token <
> ```
> 
> Not sure if there's another place I'd need to propagate some babel settings to get this working. I'm assuming this is due to .tsx

**Comment by LekoArts** [comment]
Before continuing this we need to first decide on which PR to keep. This one or https://github.com/gatsbyjs/gatsby/pull/22027

Otherwise I wouldn't want to move forward with both of them. Since the linked PR was created first it should be the preferred one and this one closed here.

**Comment by Eyas** [comment]
> Before continuing this we need to first decide on which PR to keep. This one or #22027
> 
> Otherwise I wouldn't want to move forward with both of them. Since the linked PR was created first it should be the preferred one and this one closed here.

Yup. I tried to engage on the other PR about the open questions there and what solutions I think would work. Namely, a typegen step to replace index.d.ts and a viable solution to innerRef and React's MutableRef.

**Comment by LekoArts** [comment]
Let's close this in favor of https://github.com/gatsbyjs/gatsby/pull/22027 then. Thanks for the PR though!
